token-metadata/example/src/processor.rs

```diff
@@ -0,0 +1,220 @@
+pub fn process_initialize(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: Initialize,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let metadata_info = next_account_info(account_info_iter)?;
+    let update_authority_info = next_account_info(account_info_iter)?;
+    let mint_info = next_account_info(account_info_iter)?;
+    let mint_authority_info = next_account_info(account_info_iter)?;
+
+    // scope the mint authority check, in case the mint is in the same account!
+    {
+        // IMPORTANT: this example metadata program is designed to work with any
+        // program that implements the SPL token interface, so there is no
+        // ownership check on the mint account.
+        let mint_data = mint_info.try_borrow_data()?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+
+        if !mint_authority_info.is_signer {
+            return Err(ProgramError::MissingRequiredSignature);
+        }
+        if mint.base.mint_authority.as_ref() != COption::Some(mint_authority_info.key) {
+            return Err(TokenMetadataError::IncorrectMintAuthority.into());
+        }
+    }
+
+    // get the required size, assumes that there's enough space for the entry
+    let update_authority = OptionalNonZeroPubkey::try_from(Some(*update_authority_info.key))?;
+    let token_metadata = TokenMetadata {
+        name: data.name,
+        symbol: data.symbol,
+        uri: data.uri,
+        update_authority,
+        mint: *mint_info.key,
+        ..Default::default()
+    };
+    let instance_size = get_instance_packed_len(&token_metadata)?;
+
+    // allocate a TLV entry for the space and write it in
+    let mut buffer = metadata_info.try_borrow_mut_data()?;
+    let mut state = TlvStateMut::unpack(&mut buffer)?;
+    state.alloc::<TokenMetadata>(instance_size)?;
+    state.borsh_serialize(&token_metadata)?;
+
+    Ok(())
+}
+
+/// Processes an [UpdateField](enum.TokenMetadataInstruction.html) instruction.
+pub fn process_update_field(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: UpdateField,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let metadata_info = next_account_info(account_info_iter)?;
+    let update_authority_info = next_account_info(account_info_iter)?;
+
+    // deserialize the metadata, but scope the data borrow since we'll probably
+    // realloc the account
+    let mut token_metadata = {
+        let buffer = metadata_info.try_borrow_data()?;
+        let state = TlvStateBorrowed::unpack(&buffer)?;
+        state.borsh_deserialize::<TokenMetadata>()?
+    };
+
+    check_update_authority(update_authority_info, &token_metadata.update_authority)?;
+
+    // Update the field
+    token_metadata.update(data.field, data.value);
+
+    // Update / realloc the account
+    realloc_and_borsh_serialize(metadata_info, &token_metadata)?;
```

Comment:
Curious how this doesn't also require a transfer for additional rent? Maybe I'm missing something